server/rollback_runner.go

```diff
@@ -0,0 +1,274 @@
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to extract TiDB unsupported statements from old statements %q", schema)
+	}
+	nodes, _, err := parser.New().Parse(supportStmts, "", "")
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse old statement %q", schema)
+	}
+	tableMap := make(map[string][]string)
+	for _, node := range nodes {
+		if stmt, ok := node.(*ast.CreateTableStmt); ok {
+			var columnNames []string
+			for _, col := range stmt.Cols {
+				columnNames = append(columnNames, col.Name.Name.O)
+			}
+			tableMap[stmt.Table.Name.O] = columnNames
+		}
+	}
+	return tableMap, nil
+}
+
+func (r *RollbackRunner) generateRollbackSQL(ctx context.Context, task *api.Task, payload *api.TaskDatabaseDataUpdatePayload, binlogFileNameList []string, tableMap map[string][]string) (string, error) {
+	adminDataSource := api.DataSourceFromInstanceWithType(task.Instance, api.Admin)
+	args := binlogFileNameList
+	args = append(args,
+		"--read-from-remote-server",
+		// Verify checksum binlog events.
+		"--verify-binlog-checksum",
+		// Tell mysqlbinlog to suppress the BINLOG statements for row events, which reduces the unneeded output.
+		"--base64-output=DECODE-ROWS",
+		// Reconstruct pseudo-SQL statements out of row events. This is where we parse the data changes.
+		"--verbose",
+		"--host", task.Instance.Host,
+		"--user", adminDataSource.Username,
+		// Start decoding the binary log at the log position, this option applies to the first log file named on the command line.
+		"--start-position", fmt.Sprintf("%d", payload.BinlogPosStart),
+		// Stop decoding the binary log at the log position, this option applies to the last log file named on the command line.
+		"--stop-position", fmt.Sprintf("%d", payload.BinlogPosEnd),
+	)
+	if task.Instance.Port != "" {
+		args = append(args, "--port", task.Instance.Port)
+	}
+	cmd := exec.CommandContext(ctx, mysqlutil.GetPath(mysqlutil.MySQLBinlog, common.GetResourceDir(r.server.profile.DataDir)), args...)
+	log.Debug("mysqlbinlog", zap.String("command", cmd.String()))
+	if adminDataSource.Password != "" {
+		cmd.Env = append(cmd.Env, fmt.Sprintf("MYSQL_PWD=%s", adminDataSource.Password))
+	}
+	pr, err := cmd.StdoutPipe()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create stdout pipe")
+	}
+
+	errPipe, err := cmd.StderrPipe()
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create stderr pipe")
+	}
+	defer errPipe.Close()
+	errScanner := bufio.NewScanner(errPipe)
+
+	if err := cmd.Start(); err != nil {
+		return "", errors.Wrap(err, "failed to run mysqlbinlog")
+	}
+
+	txnList, err := mysql.ParseBinlogStream(pr)
+	if err != nil {
+		return "", errors.WithMessage(err, "failed to parse binlog stream")
+	}
+
+	txnList, err = mysql.FilterBinlogTransactionsByThreadID(txnList, payload.ThreadID)
+	if err != nil {
+		return "", errors.WithMessage(err, "failed to filter binlog transactions by thread ID")
+	}
+
+	var rollbackSQLList []string
```

Comment:
This logic seems to belong to the mysql rollback implementation. And a lot of things above.